Add README's for creating perf tests (#2750)
knative/serving: test/performance/README.md

# Performance tests

Knative performance tests are tests geared towards producing useful performance
metrics of the knative system. As such they can choose to take a blackbox
point-of-view of the system and use it just like an end-user might see it. They
can also go more whiteboxy to narrow down the components under test.

> ## Adding performance tests

Reviewer: It would make sense to describe perf test infra first and then describe how to write new perf tests on top of it. 
1. Infra
2. how to write new test
3. how to view the result

Performance tests are simple go tests, that use [Fortio](#load-generator) for generating 

> load and [Tesgrid](#displaying-metrics) to see the metrics on a continous basis. For whitebox tests,

Reviewer 2: ```suggestion
load and [Tesgrid](#displaying-metrics) to see the metrics on a continuous basis. For whitebox tests,
```

> the performance tests also bring up the prometheus service alongwith the knative-serving. So, 

Reviewer 2: ```suggestion
the performance tests also bring up the prometheus service along with Knative Serving. So, 
```

you can [query prometheus](#prometheus-metrics) to get server side metrics with the load 
generator metrics.

## Load Generator

Knative has a [load generator library](https://github.com/knative/test-infra/tree/master/shared/loadgenerator)
that can be used to generate load. The load generator uses [Fortio](https://fortio.org/) for generating load based

> on the [generator options](https://github.com/knative/test-infra/blob/master/shared/loadgenerator/loadgenerator.go#L36)

Reviewer 2: I suggest not linking to specific lines, as they will quickly point to the wrong place once you update the code.

set in the test. The load generator provides the [generator results](https://github.com/knative/test-infra/blob/master/shared/loadgenerator/loadgenerator.go#L47)
which includes all the data points for the load generation request and any calculated latencies.

For eg.

```go
opts := loadgenerator.GeneratorOptions{Duration: 1*time.Minute, NumThreads: 1}
resp, err := opts.RunLoadTest(false /* resolvableDomain */)
if err != nil {
	t.Fatalf("Generating traffic via fortio failed: %v", err)
}
```

## Prometheus metrics

Knative provides a [prometheus wrapper](https://github.com/knative/test-infra/tree/master/shared/prometheus) that
provides methods to wait for prometheus to scrap for the metrics once the test is finished. It also provides a way
to query the prometheus server for any server-side metrics and then display those in [testgrid](#displaying-metrics)

For eg.
```go
promAPI, err := prometheus.PromAPI()
if err != nil {
	logger.Errorf("Cannot setup prometheus API")
}

> query := fmt.Sprintf("%s{configuration_namespace=\"%s\", configuration=\"%s\", revision=\"%s\"}", metric, test.ServingNamespace, names.Config, names.Revision)

Author: Thanks for catching this. Updating this to the actual query.

val, err := prometheus.RunQuery(context.Background(), logger, promAPI, query)
if err != nil {
	logger.Infof("Error querying metric %s: %v", metric, err)
}
```

## Displaying metrics

Knative uses [testgrid](https://testgrid.knative.dev) to show all its metrics. Performance metrics are shown on a 
separate tab for each repo like [serving](https://testgrid.knative.dev/knative-serving#performance). Once the test is
done, each test can define which metrics they want to be captured and shown on testgrid. For each metric, create a 
[testgrid testcase](https://github.com/knative/test-infra/blob/master/shared/testgrid/testgrid.go) by using the 
[CreatePerfTestCase() method](https://github.com/knative/serving/blob/master/test/performance/performance.go). 

For eg.

```go
testName := "TestPerformanceLatency"
var tc []testgrid.TestCase
for name, val := range metrics {

> 	tc = append(tc, CreatePerfTestCase(val, name), testName))

Reviewer 3: parenthesis misplaced. Also, maybe make the name of the arguments (in the code) more explicit? `tName` and `tcName` are kind of confusing. Otherwise explain the use of the parameters in the docstring.

Author: do you mean in https://github.com/knative/serving/blob/master/test/performance/performance.go#L99? I will update that in another PR. I want to keep this specific to README.

Reviewer 3: The parenthesis are still misplaced.
`CreatePerfTestCase(val, name), testName)` should be `CreatePerfTestCase(val, name, testName)`

}
```

> Once, we create all the test cases for the metrics, we can use the [CreateTestgridXML() method](https://github.com/knative/test-infra/blob/master/shared/testgrid/testgrid.go#L71)

Reviewer 2: ```suggestion
Once we create all the test cases for the metrics, we can use the [CreateTestgridXML() method](https://github.com/knative/test-infra/blob/master/shared/testgrid/testgrid.go#L71)
```

> to create the output xml that will be used as input to testgrid. If the test is run locally, it will create a file called `junit_knative.xml` and save it locally under `knative/serving/test/performance/artifacts/`. When the test 

Reviewer 3: It seems that the folder needs to be created before running the tests.

Author: Yup. Added a line indicating that this is expected. Maybe I can create that ifnot present and remove this expectation. I will try it out in a different PR and remove this expectation.
